src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -421,47 +523,120 @@ internal static void Unescape(ReadOnlySpan<byte> source, Span<byte> destination,
 
 #if BUILDING_INBOX_LIBRARY
                         var rune = new Rune(scalar);
-                        int bytesWritten = rune.EncodeToUtf8(destination.Slice(written));
+                        bool success = rune.TryEncodeToUtf8(destination.Slice(written), out int bytesWritten);
 #else
-                        EncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
+                        bool success = TryEncodeToUtf8Bytes((uint)scalar, destination.Slice(written), out int bytesWritten);
 #endif
+                        if (!success)
+                        {
+                            goto DestinationTooShort;
+                        }
+
                         Debug.Assert(bytesWritten <= 4);
                         written += bytesWritten;
-                    }
+                        break;
                 }
-                else
+
+                if (++idx == source.Length)
                 {
-                    destination[written++] = currentByte;
+                    goto Success;
+                }
+
+                if (source[idx] != JsonConstants.BackSlash)
+                {
+                    ReadOnlySpan<byte> remaining = source.Slice(idx);
+                    int nextUnescapedSegmentLength = remaining.IndexOf(JsonConstants.BackSlash);
+                    if (nextUnescapedSegmentLength < 0)
+                    {
+                        nextUnescapedSegmentLength = remaining.Length;
+                    }
+
+                    if ((uint)(written + nextUnescapedSegmentLength) >= (uint)destination.Length)
+                    {
+                        goto DestinationTooShort;
+                    }
+
+                    Debug.Assert(nextUnescapedSegmentLength > 0);
+                    switch (nextUnescapedSegmentLength)
+                    {
+                        case 1:
+                            destination[written++] = source[idx++];
+                            break;
+                        case 2:
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            break;
+                        case 3:
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            destination[written++] = source[idx++];
+                            break;
+                        default:
+                            remaining.Slice(0, nextUnescapedSegmentLength).CopyTo(destination.Slice(written));
+                            written += nextUnescapedSegmentLength;
+                            idx += nextUnescapedSegmentLength;
+                            break;
```

Comment:
why not just always `CopyTo`? How big of a difference does this make? Wouldn't it be better to make a suggestion to JIT team to expand this pattern similarly?

Reply:
Worst case scenario using `CopyTo` for a small number of bytes can result in a 20-30% regression. I'm sure `CopyTo` already does unrolling of its own, but I would guess that both `Slice` and `CopyTo` waste a few CPU cycles in bounds checking etc. that makes it relatively inefficient for short byte sequences.

Reply:
Could we make this at least a helper method with aggressive in-lining and follow up with issue for JIT team if they could improve this? It's possible they might respond with suggestion how to restructure the code to make this cheaper

Reply:
Why do need a helper method? Inlined methods are more difficult to step through.

Reply:
I'd prefer we haven't done this expansion at all but if it's absolutely necessary having this isolated at least makes it easier to see what's going on. Can we at least file an issue so that if JIT adds this improvement we can remove this?